### PR TITLE
Support CA's priority expander config in `Shoot` spec

### LINKED ISSUE #10683: Specify ClusterAutoscaler's priority in shoot CR
**How to categorize this issue?**

/area auto-scaling
/kind enhancement

**What would you like to be added**:
Currently ClusterAutoscaler's priority configmap cluster-autoscaler-priority-expander is installed by user after cluster creation.
It should be part of shoot CR. 

**Why is this needed**:
We have more than one node groups to allow system components for cluster resilience.
But the cheapest one should take priority for TCO.
But we can not do that because configmap cluster-autoscaler-priority-expander is created by us after cluster ready.
I would like to make the configmap part of shoot yaml, so that gardener controller can create it in between control plane ready and system component deployment.

### PR DESCRIPTION
**Pull request by tobschli**
**How to categorize this PR?**

/area usability
/kind api-change

**What this PR does / why we need it**:
When creating a `Shoot`, it is currently not possible to directly pass the desired priority expander configuration ([ref](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/expander/priority/readme.md)) before the first node of a shoot is ready.

**Which issue(s) this PR fixes**:
Fixes #10683 

**Special notes for your reviewer**:

**Release note**:

```user feature
It is now possible to specify a `ConfigMap` from the project namespace to be used as the `cluster-autoscaler-priority-expander` for `Shoot` clusters. This is configurable via `.spec.kubernetes.clusterAutoscaler.expanderConfig.configMap` in the `Shoot` API.
```

### CODE REVIEW

File: pkg/apis/core/types_shoot.go [hunk 1]
```diff
@@ -516,6 +519,13 @@ type ClusterAutoscaler struct {
 	Verbosity *int32
 }
 
+type ExpanderConfig struct {
+	// ConfigMapName specifies the name of the expander configuration, according to the priority expander.
+	// See: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/expander/priority/readme.md
+	// +optional
+	ConfigMapName *string `json:"configMapName,omitempty"`
```
Comment (rfranzke) [inline pkg/apis/core/types_shoot.go]: Remove this from internal APIs.
```suggestion
	ConfigMapName *string
```

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 2]
```diff
@@ -627,6 +627,9 @@ type ClusterAutoscaler struct {
 	// See: https://github.com/gardener/autoscaler/blob/machine-controller-manager-provider/cluster-autoscaler/FAQ.md#what-are-expanders.
 	// +optional
 	Expander *ExpanderMode `json:"expander,omitempty" protobuf:"bytes,7,opt,name=expander"`
+	// ExpanderConfig defines the configuration of the CA expander
+	// +optional
+	ExpanderConfig *ExpanderConfig `json:"expanderConfig,omitempty" protobuf:"bytes,15,opt,name=expanderConfig"`
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/types_shoot.go]: Can you move this to the end of the struct? It's quite confusing looking at the protobuf numbers, even if it better fits here.

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 3]
```diff
@@ -627,6 +627,9 @@ type ClusterAutoscaler struct {
 	// See: https://github.com/gardener/autoscaler/blob/machine-controller-manager-provider/cluster-autoscaler/FAQ.md#what-are-expanders.
 	// +optional
 	Expander *ExpanderMode `json:"expander,omitempty" protobuf:"bytes,7,opt,name=expander"`
+	// ExpanderConfig defines the configuration of the CA expander
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/types_shoot.go]:
```suggestion
	// ExpanderConfig defines the configuration of the CA expander.
```

File: pkg/apis/core/types_shoot.go [hunk 4]
```diff
@@ -500,6 +500,9 @@ type ClusterAutoscaler struct {
 	// Expander defines the algorithm to use during scale up (default: least-waste).
 	// See: https://github.com/gardener/autoscaler/blob/machine-controller-manager-provider/cluster-autoscaler/FAQ.md#what-are-expanders.
 	Expander *ExpanderMode
+	// ExpanderConfig defines the configuration of the CA expander
```
Comment (rfranzke) [inline pkg/apis/core/types_shoot.go]:
```suggestion
	// ExpanderConfig defines the configuration of the CA expander.
```

File: pkg/apis/core/types_shoot.go [hunk 5]
```diff
@@ -500,6 +500,9 @@ type ClusterAutoscaler struct {
...
+	// +optional
+	ExpanderConfig *ExpanderConfig `json:"expanderConfig,omitempty"`
```
Comment (rfranzke) [inline pkg/apis/core/types_shoot.go]: Remove this from internal APIs.
```suggestion
	ExpanderConfig *ExpanderConfig
```

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 6]
```diff
@@ -650,6 +653,13 @@ type ClusterAutoscaler struct {
 	Verbosity *int32 `json:"verbosity,omitempty" protobuf:"varint,14,opt,name=verbosity"`
 }
 
+type ExpanderConfig struct {
+	// ConfigMapName specifies the name of the expander configuration, according to the priority expander.
+	// See: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/expander/priority/readme.md
+	// +optional
+	ConfigMapName *string `json:"configMapName,omitempty" protobuf:"bytes,1,opt,name=configMapName"`
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/types_shoot.go]: Why is this optional? If `ExpanderConfig` is provided, I would expect that leaving it empty makes no sense?

File: pkg/admissioncontroller/webhook/auth/seed/graph/eventhandler_shoot.go [hunk 7]
```diff
@@ -136,6 +136,14 @@ func (g *graph) handleShootCreateOrUpdate(ctx context.Context, shoot *gardencore
 		g.addEdge(exposureClassVertex, shootVertex)
 	}
 
+	if configMapName := shoot.Spec.Kubernetes.ClusterAutoscaler.ExpanderConfig.ConfigMapName; configMapName != nil {
```
Comment (rfranzke) [inline pkg/admissioncontroller/webhook/auth/seed/graph/eventhandler_shoot.go]: https://github.com/gardener/gardener/blob/01391f78c30a37968fceafd885f0637e689f1909/pkg/admissioncontroller/webhook/auth/seed/graph/eventhandler_shoot.go#L45-L58 needs to be augmented, no?

File: pkg/admissioncontroller/webhook/auth/seed/graph/eventhandler_shoot.go [hunk 8]
```diff
@@ -136,6 +136,14 @@ func (g *graph) handleShootCreateOrUpdate(ctx context.Context, shoot *gardencore
...
+		if shoot.Status.SeedName != nil {
+			configMapVertex := g.getOrCreateVertex(VertexTypeConfigMap, shoot.Namespace, *configMapName)
+			seedVertex := g.getOrCreateVertex(VertexTypeSeed, "", *shoot.Status.SeedName)
+			g.addEdge(configMapVertex, seedVertex)
```
Comment (rfranzke) [inline pkg/admissioncontroller/webhook/auth/seed/graph/eventhandler_shoot.go]: I don't understand this. This is a shoot-specific `ConfigMap`. Why does it point to the seed vertex?

File: pkg/admissioncontroller/webhook/auth/seed/graph/eventhandler_shoot.go [hunk 9]
```diff
@@ -136,6 +136,14 @@ func (g *graph) handleShootCreateOrUpdate(ctx context.Context, shoot *gardencore
 		g.addEdge(exposureClassVertex, shootVertex)
 	}
 
+	if configMapName := shoot.Spec.Kubernetes.ClusterAutoscaler.ExpanderConfig.ConfigMapName; configMapName != nil {
+		if shoot.Status.SeedName != nil {
+			configMapVertex := g.getOrCreateVertex(VertexTypeConfigMap, shoot.Namespace, *configMapName)
```
Comment (rfranzke) [inline pkg/admissioncontroller/webhook/auth/seed/graph/eventhandler_shoot.go]: Augment the unit test for this new case... https://github.com/gardener/gardener/blob/master/pkg/admissioncontroller/webhook/auth/seed/graph/graph_test.go

File: pkg/admissioncontroller/webhook/auth/seed/graph/eventhandler_shoot.go (hunk 7 above)
Comment (rfranzke) [inline pkg/admissioncontroller/webhook/auth/seed/graph/eventhandler_shoot.go]: Several `nil` checks are missing for `ClusterAutoscaler` and `ExpanderConfig`

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go [hunk 10]
```diff
@@ -60,12 +62,16 @@ type Interface interface {
 	SetMachineDeployments([]extensionsv1alpha1.MachineDeployment)
 	// SetMaxNodesTotal sets the maximum number of nodes that can be created in the cluster.
 	SetMaxNodesTotal(int64)
+	// HasExpanderConfigMapConfigured returns whether a priority expander ConfigMap was configured to be used
+	HasExpanderConfigMapConfigured() bool
```
Comment (rfranzke) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: Why is this part of the interface? There are no external callers?

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go [hunk 11]
```diff
@@ -49,6 +49,8 @@ const (
 
 	portNameMetrics       = "metrics"
 	portMetrics     int32 = 8085
+
+	ClusterAutoScalerPriorityConfigMapName = "cluster-autoscaler-priority-expander"
```
Comment (rfranzke) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: Why is this exported?

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go [hunk 12]
```diff
@@ -176,6 +186,23 @@ func (c *clusterAutoscaler) Deploy(ctx context.Context) error {
 		return err
 	}
 
+	if c.HasExpanderConfigMapConfigured() {
+		configMap := &corev1.ConfigMap{}
+		if err := c.gardenClient.Get(ctx, client.ObjectKey{Name: *c.config.ExpanderConfig.ConfigMapName, Namespace: c.projectNamespace}, configMap); err != nil {
+			return err
+		}
```
Comment (rfranzke) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: Usually, we fetch such data in the `botanist` and provide the content to the component package. This way, you don't have to pass garden client or project namespace.

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go [hunk 13]
```diff
@@ -176,6 +186,23 @@ func (c *clusterAutoscaler) Deploy(ctx context.Context) error {
 		return err
 	}
 
+	if c.HasExpanderConfigMapConfigured() {
+		configMap := &corev1.ConfigMap{}
+		if err := c.gardenClient.Get(ctx, client.ObjectKey{Name: *c.config.ExpanderConfig.ConfigMapName, Namespace: c.projectNamespace}, configMap); err != nil {
+			return err
```
Comment (rfranzke) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: fmt.Errorf?

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go [hunk 14]
```diff
@@ -176,6 +186,23 @@ func (c *clusterAutoscaler) Deploy(ctx context.Context) error {
...
+
+		configMap.Namespace = c.namespace
+		configMap.Name = ClusterAutoScalerPriorityConfigMapName
+
+		if _, err := controllerutils.GetAndCreateOrMergePatch(ctx, c.client, configMap, func() error {
+			configMap.ResourceVersion = ""
+			return nil
+		}); err != nil {
+			return err
+		}
```
Comment (rfranzke) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: This way, we carry arbitrary metadata which is maybe undesired. I guess it's safer to create a new `*corev1.ConfigMap` and explicitly take over the data, and labels/annotations only.

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go [hunk 15]
```diff
@@ -176,6 +186,23 @@ func (c *clusterAutoscaler) Deploy(ctx context.Context) error {
...
+
+		configMap.Namespace = c.namespace
+		configMap.Name = ClusterAutoScalerPriorityConfigMapName
+
+		if _, err := controllerutils.GetAndCreateOrMergePatch(ctx, c.client, configMap, func() error {
+			configMap.ResourceVersion = ""
+			return nil
+		}); err != nil {
+			return err
```
Comment (rfranzke) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: fmt.Errorf?

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go [hunk 16]
```diff
@@ -176,6 +186,23 @@ func (c *clusterAutoscaler) Deploy(ctx context.Context) error {
...
+	if c.HasExpanderConfigMapConfigured() {
+		configMap := &corev1.ConfigMap{}
+		if err := c.gardenClient.Get(ctx, client.ObjectKey{Name: *c.config.ExpanderConfig.ConfigMapName, Namespace: c.projectNamespace}, configMap); err != nil {
+			return err
+		}
+
+		configMap.Namespace = c.namespace
+		configMap.Name = ClusterAutoScalerPriorityConfigMapName
+
+		if _, err := controllerutils.GetAndCreateOrMergePatch(ctx, c.client, configMap, func() error {
```
Comment (rfranzke) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: This overwrites the data on each reconciliation. I don't think this is what we want, or? IIUC, it should only be about initial creation if it does not exist?

Comment (rfranzke) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: This is the seed client, but the `ConfigMap` lives in the `kube-system` of the Shoot.

**Review by rfranzke — CHANGES_REQUESTED** [review]
You must also augment the `shoot/reference` controller (and its docs): https://github.com/gardener/gardener/tree/master/pkg/controllermanager/controller/shoot/reference

File: pkg/admissioncontroller/webhook/auth/seed/graph/eventhandler_shoot.go (hunk 8 above)
Comment (timuthy) [reply pkg/admissioncontroller/webhook/auth/seed/graph/eventhandler_shoot.go]: Please also adjust https://github.com/gardener/gardener/blob/master/docs/deployment/gardenlet_api_access.md.

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go (hunk 11 above)
Comment (tobschli) [reply pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: This is used in the respective unit test.

Comment (rfranzke) [reply pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: In unit tests, we typically avoid the usage of constants and rather inline the values. This is to ensure that we detect drift in case somebody changes the constant value.

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go (hunk 16 above)
Comment (tobschli) [reply pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: According to documentation, [the `ConfigMap` must live in the same namespace as the CA pod](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/expander/priority/readme.md#configuration).
IIUC, the CA pod gets deployed via a `Deployment`, using the seed client ([ref](https://github.com/gardener/gardener/blob/master/pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go#L179)), right?

Comment (tobschli) [reply pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: Initially, that was the idea I think, yes.
I think it could also be a good feature to let shoot owners maintain this config via the specified `ConfigMap` in the virtual-garden cluster.
Although I can imagine that this could be confusing / not wanted by cluster owners to configure this outside of the actual cluster.
If we decide to just copy the contents once during creation, I propose to rename the field in the `Shoot` spec to something like `initConfigMap` and update the documentation accordingly.
What do you think?

Comment (rfranzke) [reply pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: The namespace comes from a defaulted CLI flag: https://github.com/gardener/autoscaler/blob/45e190f4bd7890e0029bef8eb9affa0946f135d5/cluster-autoscaler/main.go#L116
We don't override it, so CA looks for the `ConfigMap` in the `kube-system` namespace of the shoot cluster. This also matches what I have observed in real systems.

File: pkg/apis/core/v1beta1/types_shoot.go (hunk 2 above)
Comment (rfranzke) [inline pkg/apis/core/v1beta1/types_shoot.go]: Augment `example/90-shoot.yaml`.

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go (hunk 16 above)
Comment (rfranzke) [reply pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: Yes, let's rename the field (proposal: `initialExpanderConfig`) and make it immutable. This way, we don't break the flow/"API" of existing users.

Comment (tobschli) [reply pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: Ah, alright! Thank you very much :)

Comment (tobschli) [reply pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: Yup, sounds good to me :)

**Comment by vlerenc** [comment]
Hi @tobschli, do you think, the feature will be ready soon? I'd like to use it immediately for seed cluster management and could already prepare the necessary changes.

**Comment by tobschli** [comment]
> Hi @tobschli, do you think, the feature will be ready soon? I'd like to use it immediately for seed cluster management and could already prepare the necessary changes.

I can continue on this today, although I cannot promise that I get it ready by today.
Will do my best tho!

**Comment by tobschli** [comment]
I have implemented the changes mentioned in the review, as well as what we talked about today @rfranzke.
However, there are still some parts in my implementation that I am not really happy with, especially the part that I marked with `TODO` in the latest commit.
Do you have any ideas how I can improve here?

File: pkg/admissioncontroller/webhook/auth/seed/graph/eventhandler_shoot.go [hunk 17]
```diff
@@ -136,6 +137,18 @@ func (g *graph) handleShootCreateOrUpdate(ctx context.Context, shoot *gardencore
...
+	if caConfig := shoot.Spec.Kubernetes.ClusterAutoscaler; caConfig != nil {
+		if expanderConfig := caConfig.ExpanderConfig; expanderConfig != nil {
+			if configMapName := expanderConfig.ConfigMapName; len(configMapName) != 0 {
+				if shoot.Status.SeedName != nil {
+					configMapVertex := g.getOrCreateVertex(VertexTypeConfigMap, shoot.Namespace, configMapName)
+					g.addEdge(configMapVertex, shootVertex)
+				}
+			}
+		}
+	}
```
Comment (rfranzke) [inline pkg/admissioncontroller/webhook/auth/seed/graph/eventhandler_shoot.go:150]: Well, why not simply like this? You should ensure via validation that the `ConfigMapName` has `length > 0` when provided.

```suggestion
	if shoot.Spec.Kubernetes.ClusterAutoscaler != nil && shoot.Spec.Kubernetes.ClusterAutoscaler.ExpanderConfig != nil {
		configMapVertex := g.getOrCreateVertex(VertexTypeConfigMap, shoot.Namespace, configMapName)
		g.addEdge(configMapVertex, shootVertex)
	}
```

I don't understand the `shoot.Status.SeedName != nil` condition. Can you elaborate?

File: pkg/admissioncontroller/webhook/auth/seed/graph/graph_test.go [hunk 18]
```diff
@@ -71,6 +71,7 @@ var _ = Describe("graph", func() {
 		shootIssuerNamespace = "gardener-system-shoot-issuer"
 
 		shoot1                           *gardencorev1beta1.Shoot
+		shoot1ExpanderConfigMap          *corev1.ConfigMap
```
Comment (rfranzke) [inline pkg/admissioncontroller/webhook/auth/seed/graph/graph_test.go:74]:
```suggestion
		shoot1CAExpanderConfigMap          *corev1.ConfigMap
```

File: pkg/admissioncontroller/webhook/auth/seed/graph/graph_test.go [hunk 19]
```diff
@@ -213,6 +225,11 @@ var _ = Describe("graph", func() {
 					Providers: []gardencorev1beta1.DNSProvider{shoot1DNSProvider1, shoot1DNSProvider2},
 				},
 				Kubernetes: gardencorev1beta1.Kubernetes{
+					ClusterAutoscaler: &gardencorev1beta1.ClusterAutoscaler{
+						ExpanderConfig: &gardencorev1beta1.ExpanderConfig{
+							ConfigMapName: "expanderConfig",
```
Comment (rfranzke) [inline pkg/admissioncontroller/webhook/auth/seed/graph/graph_test.go:230]:
```suggestion
							ConfigMapName: shoot1ExpanderConfigMap.Name,
```

File: pkg/apis/core/types_shoot.go [hunk 20]
```diff
@@ -500,6 +500,9 @@ type ClusterAutoscaler struct {
 	// Expander defines the algorithm to use during scale up (default: least-waste).
 	// See: https://github.com/gardener/autoscaler/blob/machine-controller-manager-provider/cluster-autoscaler/FAQ.md#what-are-expanders.
 	Expander *ExpanderMode
+	// ExpanderConfig defines the configuration of the CA expander.
+	// +optional
```
Comment (rfranzke) [inline pkg/apis/core/types_shoot.go:504]:
```suggestion
```

Not needed in internal API

File: pkg/apis/core/types_shoot.go [hunk 21]
```diff
@@ -500,6 +500,9 @@ type ClusterAutoscaler struct {
 	// Expander defines the algorithm to use during scale up (default: least-waste).
 	// See: https://github.com/gardener/autoscaler/blob/machine-controller-manager-provider/cluster-autoscaler/FAQ.md#what-are-expanders.
 	Expander *ExpanderMode
+	// ExpanderConfig defines the configuration of the CA expander.
```
Comment (rfranzke) [inline pkg/apis/core/types_shoot.go:503]: Nit: This doc string isn't quite helpful - what is the CA expander/what is it used for?

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 22]
```diff
@@ -648,6 +648,16 @@ type ClusterAutoscaler struct {
 	// Verbosity allows CA to modify its log level (default: 2).
 	// +optional
 	Verbosity *int32 `json:"verbosity,omitempty" protobuf:"varint,14,opt,name=verbosity"`
+	// ExpanderConfig defines the configuration of the CA expander.
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/types_shoot.go:651]: Nit: Same here

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 23]
```diff
@@ -648,6 +648,16 @@ type ClusterAutoscaler struct {
...
+	// +optional
+	ExpanderConfig *ExpanderConfig `json:"expanderConfig,omitempty" protobuf:"bytes,15,opt,name=expanderConfig"`
+}
+
+// ExpanderConfig contains the configuration for the CA expander.
+type ExpanderConfig struct {
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/types_shoot.go:657]:
```suggestion
type ClusterAutoscalerExpanderConfig struct {
```

to make it clear

File: pkg/apis/core/types_shoot.go [hunk 24]
```diff
@@ -516,6 +519,13 @@ type ClusterAutoscaler struct {
 	Verbosity *int32
 }
 
+// ExpanderConfig contains the configuration for the CA expander.
+type ExpanderConfig struct {
```
Comment (rfranzke) [inline pkg/apis/core/types_shoot.go:523]:
```suggestion
type ClusterAutoscalerExpanderConfig struct {
```

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 25]
```diff
@@ -648,6 +648,16 @@ type ClusterAutoscaler struct {
 	// Verbosity allows CA to modify its log level (default: 2).
 	// +optional
 	Verbosity *int32 `json:"verbosity,omitempty" protobuf:"varint,14,opt,name=verbosity"`
+	// ExpanderConfig defines the configuration of the CA expander.
+	// +optional
+	ExpanderConfig *ExpanderConfig `json:"expanderConfig,omitempty" protobuf:"bytes,15,opt,name=expanderConfig"`
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/types_shoot.go:653]: Adapt `example/90-shoot.yaml`.

Generally, please revisit the comments of the previous PR reviews - some things are still missing.

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go [hunk 26]
```diff
@@ -70,30 +72,33 @@ func New(
...
-	config         *gardencorev1beta1.ClusterAutoscaler
-	maxNodesTotal  int64
-	runtimeVersion *semver.Version
+	client            client.Client
+	namespace         string
+	secretsManager    secretsmanager.Interface
+	image             string
+	replicas          int32
+	config            *gardencorev1beta1.ClusterAutoscaler
+	expanderConfigMap *corev1.ConfigMap
```
Comment (rfranzke) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go:99]: I think it's enough to just pass the real configuration:

```suggestion
	expanderConfiguration *string
```

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go [hunk 27]
```diff
@@ -385,6 +390,13 @@ func (c *clusterAutoscaler) SetMaxNodesTotal(maxNodesTotal int64) {
 	c.maxNodesTotal = maxNodesTotal
 }
 
+func (c *clusterAutoscaler) hasExpanderConfigMapConfigured() bool {
+	if c.config == nil || c.config.ExpanderConfig == nil {
+		return false
+	}
+	return len(c.config.ExpanderConfig.ConfigMapName) != 0
+}
+
```
Comment (rfranzke) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go:399]: This func is not needed anymore

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go [hunk 28]
```diff
@@ -479,8 +491,7 @@ func (c *clusterAutoscaler) computeCommand() []string {
 
 func (c *clusterAutoscaler) computeShootResourcesData(serviceAccountName string) (map[string][]byte, error) {
 	var (
-		registry = managedresources.NewRegistry(kubernetes.ShootScheme, kubernetes.ShootCodec, kubernetes.ShootSerializer)
-
```
Comment (rfranzke) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go:483]: Why this change? Revert it

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go [hunk 29]
```diff
@@ -610,10 +621,27 @@ func (c *clusterAutoscaler) computeShootResourcesData(serviceAccountName string)
...
 
-	return registry.AddAllAndSerialize(
-		clusterRole,
-		clusterRoleBinding,
-		role,
-		rolebinding,
-	)
+	objects := []client.Object{clusterRole, clusterRoleBinding, role, rolebinding}
+
+	if c.hasExpanderConfigMapConfigured() {
```
Comment (rfranzke) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go:626]:
```suggestion
	if c.expanderConfiguration != nil {
```

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go [hunk 30]
```diff
@@ -610,10 +621,27 @@ func (c *clusterAutoscaler) computeShootResourcesData(serviceAccountName string)
...
+		annotations := map[string]string{}
+		if c.expanderConfigMap.Annotations != nil {
+			annotations = c.expanderConfigMap.Annotations
+		}
+		annotations[resourcesv1alpha1.Ignore] = "true"
```
Comment (rfranzke) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go:631]: This complexity is not needed - just embed the annotations below in L638

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go [hunk 31]
```diff
@@ -49,6 +49,8 @@ const (
 
 	portNameMetrics       = "metrics"
 	portMetrics     int32 = 8085
+
+	clusterAutoScalerPriorityConfigMapName = "cluster-autoscaler-priority-expander"
```
Comment (rfranzke) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go:53]: Since this name is only used at a single place, I would remove this const and inline the name where needed.

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go [hunk 32]
```diff
@@ -610,10 +621,27 @@ func (c *clusterAutoscaler) computeShootResourcesData(serviceAccountName string)
...
+		annotations[resourcesv1alpha1.Ignore] = "true"
+
+		expanderConfigMap := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:        clusterAutoScalerPriorityConfigMapName,
+				Namespace:   metav1.NamespaceSystem,
+				Labels:      c.expanderConfigMap.Labels,
+				Annotations: annotations,
+			},
+			Data: c.expanderConfigMap.Data,
```
Comment (rfranzke) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go:640]: I don't think we should copy all data - just the one necessary for CA. 
```suggestion
			Data: map[string]string{"priorities": *c.expanderConfiguration},
```

File: pkg/gardenlet/operation/botanist/clusterautoscaler.go [hunk 33]
```diff
@@ -22,19 +23,28 @@ import (
...
 	}
 
+	var expanderConfigMap *corev1.ConfigMap = nil
+	if b.Shoot.GetInfo().Spec.Kubernetes.ClusterAutoscaler != nil &&
+		b.Shoot.GetInfo().Spec.Kubernetes.ClusterAutoscaler.ExpanderConfig != nil {
+		err = b.GardenClient.Get(ctx, client.ObjectKey{Name: b.Shoot.GetInfo().Spec.Kubernetes.ClusterAutoscaler.ExpanderConfig.ConfigMapName, Namespace: b.Shoot.GetInfo().Namespace}, expanderConfigMap)
+		if err != nil {
+			return nil, err
+		}
+	}
```
Comment (rfranzke) [inline pkg/gardenlet/operation/botanist/clusterautoscaler.go:39]:
```suggestion
	var expanderConfiguration *string
	if b.Shoot.GetInfo().Spec.Kubernetes.ClusterAutoscaler != nil &&
		b.Shoot.GetInfo().Spec.Kubernetes.ClusterAutoscaler.ExpanderConfig != nil {
		configMap := &corev1.ConfigMap{ObjectMeta: metav1.ObjectMeta{Name: b.Shoot.GetInfo().Spec.Kubernetes.ClusterAutoscaler.ExpanderConfig.ConfigMapName, Namespace: b.Shoot.GetInfo().Namespace}}
		if err := b.GardenClient.Get(ctx, client.ObjectKeyFromObject(configMap), configMap); err != nil {
			return nil, fmt.Errorf("failed reading expander configuration from ConfigMap %s referenced in shoot: %w", client.ObjectKeyFromObject(configMap), err)
		}
		// TODO: This `priority` data key should probably be a constant in the `clusterautoscaler` component package and get referenced from there.
		expanderConfiguration = configMap.Data["priorities"]
	}

```

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: pkg/gardenlet/operation/botanist/clusterautoscaler.go (hunk 33 above)
Comment (vlerenc) [reply pkg/gardenlet/operation/botanist/clusterautoscaler.go:39]: "priorities" or "priority"?

Comment (rfranzke) [reply pkg/gardenlet/operation/botanist/clusterautoscaler.go:39]: Thanks, changed :)

**Comment by tobschli** [comment]
/close
In favor of implementation mentioned in [this comment](https://github.com/gardener/gardener/issues/10683#issuecomment-2530910388)
